# Day 1 Reading Journal

Hi everyone.  These are the exercises for the first set of readings from ThinkPython.  While this journal consists of several required exercises, it is meant to encourage active reading more generally.  You should use this journal to take detailed notes, catalog questions, and explore the content from ThinkPython deeply.

For more information on using the Jupyter notebook, check out the documentation [here](https://www.datacamp.com/community/tutorials/tutorial-jupyter-notebook##UseJupyter).

Additionally, in order to write text in this document, you will be making use of Markdown cells. A handy cheatsheet for Markdown is available [here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet).

## Notes

types - strings, integers, floats
  
  numbers with decimals are floats. Can convert types
  
comment with #

state diagrams - graphical representations of code

To define a function

  def "function name" :
    don't forget the colon
    
import math for math functions

input function = input("input prompt")

Variables defined in the function aren't global

Modulus can be used to find remainders

Floor and ceiling functions can be used wit the math module to round up and down.

### Recursion

Functions can call themself.

You can pass a parameter in the argument that defines how many times the function is recursive for.

Recursion can happen infinitely but is generally bad.

## Chapter 1

Note: the exercise numbers below match up with the reading for ease of cross referencing.

### Exercise 1.4

Start the Python interpreter and use it as a simple calculator. Python’s syntax for math operations is almost the same as standard mathematical notation. For example, the symbols +, - and / denote addition, subtraction and division, as you would expect. The symbol for multiplication is *

* How many seconds are there in 42 minutes 42 seconds?
* How many miles are there in 10 kilometers? Hint: there are 1.61 kilometers in a mile.
* If you run a 10 kilometer race in 42 minutes 42 seconds, what is your average pace (time per mile in minutes and seconds)? What is your average speed in miles per hour?

Please use the following markdown cell to provide your answers.

In [1]:
10/1.61/0.7116666666666667

8.727653570337614

##### Question 1
2562 seconds
##### Question 2
6.21 miles
##### Question 3
26 minutes 31 seconds pace

8.73 mph




## Chapter 2

### Exercise

Assume that we execute the following assignment statements:

```
width = 17
height = 12.0
delimiter = '.'
```

For each of the following expressions, write the value of the expression and the type (of the value of the expression).

Example: `width * 2` <br/>
Answer: 34, integer

1. `width/2`
2. `width/2.0`
3. `height/3`
4. `1 + 2 * 5`
5. `delimiter * 5`

Please use the following markdown cell to provide your answer.

In [2]:
width = 17
height = 12.0
delimiter = '.'

delimiter*5


'.....'

##### Question 1
width/2

8.5, float

##### Question 2
width/2.0

8.5, float

##### Question 3
height/3

4.0, float

##### Question 4
1+2*5

11, integer

##### Question 5
delimiter*5

'.....' , string

### Exercise 2.2

Practice using the Python interpreter as a calculator.  For each of these provide Python code that computes each answer.

The volume of a sphere with radius r is $4/3 \pi  r^3$. What is the volume of a sphere with radius 5? <br />
Hint: 392.7 is wrong! You can approximate pi with 3.14 or import it through the Python math library.

In [7]:
value = (4/3) * 3.14 * (5^3)
value = round(value, 2)
print(value)

25.12


Suppose the cover price of a book is \$24.95, but bookstores get a 40% discount. Shipping costs \$3 for the first copy and 75 cents for each additional copy. What is the total wholesale cost for 60 copies?

In [25]:
Price = 24.95
Discounted = 24.95*.60
Books = 60
BCost = Discounted*Books

Shipping = 3 + (Books-1)*.75
TotalCost = BCost + Shipping

print(round(TotalCost,2))

945.45


If I leave my house at 6:52 am and run 1 mile at an easy pace (8:15 per mile), then 3 miles at tempo (7:12 per mile) and 1 mile at easy pace again, what time do I get home for breakfast? 

In [4]:
HoursNow = 6
MinutesNow = 52
MinutesNow = HoursNow*60 + MinutesNow

Pace1m = 8
Pace1s = 15
Pace1 = Pace1m + Pace1s/60

Pace1distance = 2
Pace1time = Pace1*Pace1distance

Pace2m = 7
Pace2s = 12
Pace2 = Pace2m+Pace2s/60

Pace2distance = 3
Pace2time = Pace2*Pace2distance

TotalM = MinutesNow+Pace1time+Pace2time
Decimal = TotalM/60

import math
H = math.floor(Decimal)
M = round((Decimal - H)*60)

Hs = str(H)
Ms = str(M)

Time = Hs +":"+ Ms+" am"
print(Time)

7:30 am


## Chapter 3

### Exercise 3.1

Write a function named `right_justify` that takes a string named s as a parameter and prints the string with enough leading spaces so that the last letter of the string is in column 70 of the display.

```
>>> right_justify('monty')
                                                                 monty
```
Hint: Use string concatenation and repetition. Also, Python provides a built-in function called `len` that returns the length of a string, so the value of `len('monty')` is 5.

In [17]:
def right_justify(s):
    length = len(s)
    space = " "
    spaces = 70 - length
    
    indent = spaces*space
    justify = indent + s
    
    print(justify)

right_justify('I think it worked')
right_justify('double check')

                                                     I think it worked
                                                          double check


### Exercise 3.3

Note: This exercise should be done using only the statements and other features we have learned so far.

1\. Write a function that draws a grid like the following:

```
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
```
Note: There are four empty spaces between the vertical bars.

To print more than one value on a line, you can print a comma-separated sequence of values:

```
print('+', '-')
```

By default, `print` advances to the next line, but you can override that behavior and put a space at the end, like this:

```
print('+', end=' ')
print('-')
```

The output of these statements is '+ -'.

A `print` statement with no argument ends the current line and goes to the next line.

In [13]:
def grid():
    plusline = "+ - - - - + - - - - + - - - - + - - - - +"
    spacelin = "|         |         |         |         |"
    
    
    print(plusline)
    print(spacelin)
    print(spacelin)
    print(spacelin)
    print(spacelin)
    print(plusline)
    print(spacelin)
    print(spacelin)
    print(spacelin)
    print(spacelin)
    print(plusline)
    print(spacelin)
    print(spacelin)
    print(spacelin)
    print(spacelin)
    print(plusline)
    print(spacelin)
    print(spacelin)
    print(spacelin)
    print(spacelin)
    print(plusline)
    
grid()

+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +


2\. Write a function that draws a similar grid with four rows and four columns.

## Chapter 5.1 - 5.7

### Exercise 5.2

Fermat’s Last Theorem says that there are no positive integers $a$, $b$, and $c$ such that $$a^n + b^n = c^n$$ for any values of $n$ greater than 2.

1\. Write a function named `check_fermat` that takes four parameters – $a$, $b$, $c$ and $n$ – and that checks to see if Fermat’s theorem holds. If n is greater than 2 and it turns out to be true that
$a^n + b^n = c^n$, then the program should print: "Holy smokes, Fermat was wrong!" Otherwise the program should print: "No, that doesn’t work."

In [ ]:
def check_fermat(a,b,c,n):
    if a^n + b^n == c^n:
        print("Holy smokes, Fermat was wrong!")
    else:
        print("No, that doesn't work")
        
check_fermat(2,4,10,2)

2\. Write a function that prompts the user to input values for $a$, $b$, $c$ and $n$, converts them to integers, and uses `check_fermat` to check whether they violate Fermat’s theorem.

In [3]:
def interactive_fermat():
    a = int(input("a"))
    b = int(input("b"))
    c = int(input("c"))
    n = int(input("n"))
    
    if a^n + b^n == c^n:
        print("Holy smokes, Fermat was wrong!")
    else:
        print("No, that doesn't work")


interactive_fermat()

a3
b2
c6
n4
No, that doesn't work


### Exercise 5.3 

If you are given three sticks, you may or may not be able to arrange them in a triangle. For example, if one of the sticks is 12 inches long and the other two are one inch long, it is clear that you will not be able to get the short sticks to meet in the middle. For any three lengths, there is a simple test to see if it is possible to form a triangle:
> If any of the three lengths is greater than the sum of the other two, then you cannot form a triangle. Otherwise, you can. (If the sum of two lengths equals the third, they form what is called a “degenerate” triangle.)

1\. Write a function named `is_triangle` that takes three integers as arguments, and that prints either "Yes" or "No," depending on whether you can or cannot form a triangle from sticks with the given lengths.

In [12]:
def is_triangle(stick1,stick2,stick3):
      
    if stick1 + stick2 < stick3:
        print("no")
    elif stick2+stick3 < stick1:
        print("no")
    elif stick3+stick1 < stick2:
        print("no")
    else: 
        print("yes")
        
is_triangle(23,3,5)
    

no


2\. Write a function that prompts the user to input three stick lengths, converts them to integers, and uses `is_triangle` to check whether sticks with the given lengths can form a triangle.

In [ ]:
def is_triangle():
    
    stick1 = int(input("stick 1 length"))
    stick2 = int(input("stick 2 length"))
    stick3 = int(input("stick 3 length"))
    
    if stick1 + stick2 < stick3:
        print("no")
    elif stick2+stick3 < stick1:
        print("no")
    elif stick3+stick1 < stick2:
        print("no")
    else: 
        print("yes")
        
is_triangle()

## Quick poll
About how long did you spend working on this Reading Journal?

About 2 hours

## Notes for the Instructors

Please use the space below to write comments to help us plan the next class session.  For instance, if you want to see us go over an example of a particular type of problem, you can indicate that here.

Please remember that the best way to get quick feedback from the instructors as well as your peers, is to use Piazza.  However, consider using this channel if it doesn't make sense to post your note to Piazza.